bids/analysis/base.py

```diff
@@ -0,0 +1,214 @@
+        # subset the data if needed
+        data = self.design_matrix
+        if selectors:
+            # TODO: make sure this handles constraints on int columns properly
+            bad_keys = list(set(selectors.keys()) - set(data.columns))
+            if bad_keys:
+                raise ValueError("The following query constraints do not map "
+                                 "onto existing columns: %s." % bad_keys)
+            query = ' and '.join(["{} in {}".format(k, v)
+                                  for k, v in selectors.items()])
+            data = data.query(query)
+        return data
+
+    def get_contrast_matrix(self):
+        pass
+
+    def _drop_columns(self, data):
+        entities = {'onset', 'duration', 'run', 'session', 'subject', 'task'}
+        common_ents = list(entities & set(data.columns))
+        return data.drop(common_ents, axis=1)
+
+    def _get_groupby_cols(self, level):
+        # Get a list of keywords that define the grouping at the current level.
+        # Note that we need to include all entities *above* the current one--
+        # e.g., if the block level is 'run', this means we actually want to
+        # groupby(['subject', 'session', 'run']), otherwise we would only
+        # end up with n(runs) groups.
+        if level is None:
+            return None
+        hierarchy = ['subject', 'session', 'run']
+        pos = hierarchy.index(level)
+        return hierarchy[:(pos + 1)]
+
+    def apply_transformations(self):
+        ''' Apply all transformations to the variables in the manager.
+        '''
+        for t in self.transformations:
+            kwargs = dict(t)
+            func = kwargs.pop('name')
+            cols = kwargs.pop('input', None)
+
+            if isinstance(func, string_types):
+                if not hasattr(transform, func):
+                    raise ValueError("No transformation '%s' found!" % func)
+                func = getattr(transform, func)
+                func(self.manager, cols, **kwargs)
+
+    def get_Xy(self, **selectors):
+        ''' Return X and y information for all groups defined by the current
+        level.
+        Args:
+            selectors (dict): Optional keyword arguments to further constrain
+                the data retrieved.
+        Returns:
+            A list of triples, where each triple contains data for a single
+                group, and the elements reflect (in order):
+                    - The design matrix containing all of the predictors (X);
+                    - The filename of the 4D image associated with the
+                      current group/design matrix;
+                    - A dict of entities defining the current group (e.g.,
+                      {'subject': '03', 'run': 1})
+        '''
+        data = self._get_design_matrix(**selectors)
+        ent_cols = self._get_groupby_cols(self.level)
+
+        tuples = []
+        ent_cols = list(set(ent_cols) & set(data.columns))
```

Comment:
Is this ever not empty? `ent_cols` is a subset of `{'subject', 'session', 'run'}`. I'm not sure how to wrangle a design matrix with any of these columns in it. For instance, for the `model.json` I'm working with, I get:

```Python
>>> list(data.columns)
['intercept',
 'weight',
 'trial_type_Finger',
 'trial_type_Foot',
 'trial_type_Lips',
 'task']
```

Reply:
It's definitely not empty, otherwise the block immediately after wouldn't work (and nothing would be returned). But it's only used internally; by the time you get the data back (in the list of triples), the entities are stored separately (in the third element of the tuple), and not included in the design matrix itself (so that you don't need to worry about stripping them out yourself).

Reply:
We could probably add an argument for `drop_entities=True` though, in the event that the user wants to actually operate on the entity columns as part of the design matrix. That's how `get_design_matrix` is set up in the `BIDSVariableManager`, so I don't think there would be any harm in replicating that behavior in `get_Xy`.